### ISSUE TITLE
ecsv files fail to write with units in metadata

### ISSUE THREAD

**Issue by lauralwatkins** [astropy/astropy#5604]
This code fails because of the units in the metadata:
```
from astropy import table, units as u
t = table.Table()
t.meta = {"foo": 5*u.km, "foo2": u.s}
t["bar"] = [7]*u.km
t.write("test.dat", format="ascii.ecsv")
```
Note that either a quantity or a unit will cause the failure.

This is the end of the error messages I'm getting:
```
~/anaconda3/envs/py2/lib/python2.7/site-packages/yaml/representer.pyc in represent_object(self, data)
    408             reduce = copy_reg.dispatch_table[cls](data)
    409         elif hasattr(data, '__reduce_ex__'):
--> 410             reduce = data.__reduce_ex__(2)
    411         elif hasattr(data, '__reduce__'):
    412             reduce = data.__reduce__()

~/anaconda3/envs/py2/lib/python2.7/copy_reg.pyc in _reduce_ex(self, proto)
     68     else:
     69         if base is self.__class__:
---> 70             raise TypeError, "can't pickle %s objects" % base.__name__
     71         state = base(self)
     72     args = (self.__class__, base, state)

TypeError: can't pickle int objects
```

**Comment by lauralwatkins** [comment]
Also, @eteq had me test this using the v1.3rc1 and the problem persisted.

**Comment by eteq** [comment]
I see the same on 1.2.1... but @taldcroft, I thought this was part of what the new YAML machinery was supposed to fix?

**Comment by mhvk** [comment]
I don't think #5486 went beyond making a yaml dumper, i.e., the machinery is not yet used in `ascii.ecsv`. This would obviously be a nice extension, but perhaps @taldcroft wanted to be sure bugs would be shaken out?

**Comment by mhvk** [comment]
Checking code, I can get your example to work trivially by indeed using the `AstropyDumper` and `AstropyLoader` -- see https://github.com/astropy/astropy/compare/master...mhvk:io-ecsv-use-new-yaml-dumper?expand=1.

This passes the `table` and `io` tests and may well be what @taldcroft had in mind... But he should check, and it obviously needs test cases, etc. (for which I won't have time... feel free to proceed with the above branch).

**Comment by taldcroft** [comment]
Integration with ecsv was indeed the intent!  I was thinking about full integration with full support for all mixin columns.  ECSV uses its own custom dumper for columns and ordered dict so that needs to be considered. Offhand I am surprised that @mhvk's experiment passed tests.

**Comment by taldcroft** [comment]
I likely won't have time today (niece's graduation, full day with family), but it would be great if someone could drive this home for 1.3 as a bug fix (otherwise 1.3.1).  Basically just need a test writing and reading back a single ECSV table that has a unit, quantity, skycoord, and time object in the `meta`.  I believe this is sufficient due to solid coverage for the base machinery of ECSV and YAML serialization.

**Comment by taldcroft** [comment]
OK, I snuck away long enough to make #5605.
